rfcs/20221006-frontend_event_tracking.md

```diff
@@ -0,0 +1,294 @@
+    mediaType: MediaType;
+  };
+  /** The user uses the 'back to search' link on a single result **/
+  BACK_TO_SEARCH: {
+    /** The unique ID of the media */
+    id: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The visibility of the filter sidebar on desktop is toggled **/
+  TOGGLE_FILTERS: {
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+};
+type AnalyticsEventTypes = keyof AnalyticsEvents;
+
+/**
+ * Send an event to our analytics endpoint. Should be a no-op if the
+ * user has opted-out of analytics.
+ */
+async function sendEvent<T extends AnalyticsEventTypes>(
+  key: T,
+  payload?: AnalyticsEvents[T]
+): Promise<any> {
+  /**
+   * The default values to send with every analytics event.
+   */
+  const defaults = {
+    timestamp: Date.now(),
+    path: window.location.pathname,
+    // The dimensions of the user's device
+    width: window.innerWidth,
+    height: window.innerHeight,
+    referrer: document.referrer || null,
+  };
+  /**
+   * Actually send the event here, with some additional values.
+   * I would have prefered to use `navigator.sendBeacon` here
+   * but it is blocked by most ad blockers.
+   *
+   * We might not want to call our api route "analytics" to
+   * further avoid blocking by adblockers, but I'll use that
+   * name for clarity here.
+   */
+  const response = await fetch(`/analytics`, {
+    method: "POST",
+    headers: { "Content-Type": "application/json" },
+    body: JSON.stringify({ ...defaults, ...payload }),
+  });
+}
+```
+
+### Implementation Plan
+
+This needs more elaboration, as it's a very rough outline. For example, specific events will require more detailed planning.
+
+- [ ] Create "Event Tracking" GitHub milestone in the frontend repository
+- [ ] Design: design a privacy and analytics opt-out checkbox on the Privacy page
+- [ ] Create an issue for the `ANALYTICS_ENABLED` feature flag
+- [ ] Create an issue for the creation of the 'sendEvent' function and for the api route to post analytics events to
+- [ ] Create a single issue for each event in the sample implementation. Each issue will need its specific implementation (when the event is called, what user DOM interaction triggers it, etc.) approved by frontend developers before it is implemented.
+
+## Concerns / Pitfalls
+
+- Related to content safety, this is an area where we are surfacing user information in the form of text, or even malicious URLs, to Openverse developers. These searches could include explicit or sensitive material. We're also potentially surfacing explicit image results to anyone working with this data.
+- I think it's great to build this decoupled from a specific back-end, but if we never actually implement a backend this would largely be a wasted effort.
+- We're building something here that is commonly drop-in when 3rd party tools are used. At least for page views; custom events still require custom work. Is this a poor use of time?
```

Comment:
I think my main concern with this approach is that we would not be able to easily track a series of events for a particular user session, unless Plausible or the other backend somehow does this for us. Though if we're proxying the events through the Nuxt server, then cookie based approaches for that on the analytics tool end of things probably will not work.

This would be useful for example if we wanted to analyse what behaviours people follow when they fall off the funnel and measure A -> B -> C type funnels. For example, under the current proposal, would we be able to know if someone started from the homepage, entered a search, refined the search, and then landed on a particular result, then clicked to visit the upstream provider? To do that we'd need to be able to tie events to each other through some kind of signifier. The signifier would need to be "sticky" so that if someone opened a new tab on the search results page (opening a single result in the new tab) then we'd still be able to tell if that search session led to a click through to the provider or attribution copy event, for example (and crucially distinguish it from someone landing directly on a single result page from an unrelated link).

My understanding is that when you use 3rd party scripts directly, these sorts of things are automatically tracked together for you.

Reply:
Oh, that's actually an accidental omission on my part and i will amend the RFC. All of the drop-in sessions do a bit of work to generate a session id, typically cookie-based. We could do that as well and pass in the session id as a default with every event.

Reply:
Oh and in the case of Plausible, they actually take the user's IP and UA string, hash it with a rotating salt, to detect users and sessions. Details here:

https://plausible.io/data-policy#how-we-count-unique-users-without-cookies

So in that case we would just send requests to the self-hosted version of Plausible the correct headers and it should work automatically. We'd post our events to this endpoint: https://plausible.io/docs/events-api

Reply:
GoatCounter's design doc for session ids is really great too: 
https://github.com/arp242/goatcounter/blob/master/docs/sessions.markdown

Reply:
@zackkrida would recording the user's IP be considered invasive, even if we  rotate it daily?

Reply:
> We could also look at a custom event for a given user and all previous VIEW_PAGE events, something like:

What does this mean? In case you meant that we would change the event name for each user (like by prefixing some user hash to the event name or something along those lines), I don't think that's possible or reliable, at least. From what I understand a lot of these tools are not optimised for the event name to vary that often (per user) and they should be stable. Sometimes you even have to configure the event and its payload specifically in the tool first before they even get ingested (GTM is like this, no experience with Plausible to say either way).

Reply:
I've added the fields necessary for session and funnel tracking, along with details about security, in 34fea2a57093c38c25ae218a8fc4ac197d3bea6d. As-written we're collecting enough information to be compatible with many 3rd party services. Following the scope of this PR I'll leave the specific implementation of session id generation and the storing of event payloads to the backend service.